service/src/main/java/bio/terra/tanagra/service/AnnotationService.java

```diff
@@ -138,10 +163,119 @@ public AnnotationValue updateAnnotationValue(
+
+    // Traverse through sorted values and update table-to-be-returned. By the time we're done, we
+    // will have latest value for each table cell.
+    Table<String, String, String> tableToReturn = HashBasedTable.create();
+    sortedValuesForAllReviews.forEach(
+        value -> {
+          Annotation annotation =
+              getAnnotation(studyId, cohortRevisionGroupId, value.getAnnotationId());
+          tableToReturn.put(
+              value.getEntityInstanceId(), // row
+              annotation.getDisplayName(), // column
+              value.getLiteral().toString() // value
+              );
+        });
+    return tableToReturn;
+  }
+
+  /**
+   * @param values is a table where row is entity instance id, column is annotation name, value is
+   *     annotation value
+   * @return GCS signed URL of GCS file containing annotation values CSV. Columns are: entity
+   *     instance id, all annotation names. Cells contain annotation values.
+   */
+  public String writeAnnotationValuesToGcs(
+      String studyId, String cohortId, Table<String, String, String> values) {
+    // Convert table of annotation values to String representing CSV file
+    Cohort cohort = cohortService.getCohort(studyId, cohortId);
+    Underlay underlay = underlaysService.getUnderlay(cohort.getUnderlayName());
+    String primaryIdSourceColumnName =
+        underlay
+            .getPrimaryEntity()
+            .getIdAttribute()
+            .getMapping(MappingType.SOURCE)
+            .getValue()
+            .getColumnName();
+    StringBuilder columnHeaders = new StringBuilder(primaryIdSourceColumnName);
+    List<Annotation> annotations =
+        getAllAnnotations(studyId, cohortId, /*offset=*/ 0, /*limit=*/ Integer.MAX_VALUE);
+    annotations.forEach(
+        annotation -> {
+          columnHeaders.append(String.format(",%s", annotation.getDisplayName()));
+        });
+    StringBuilder fileContents = new StringBuilder(columnHeaders + "\n");
+    values
+        .rowKeySet()
+        .forEach(
+            entityInstanceId -> {
+              StringBuilder row = new StringBuilder(entityInstanceId);
+              annotations.forEach(
+                  annotation -> {
+                    String value =
+                        values.contains(entityInstanceId, annotation.getDisplayName())
+                            ? values.get(entityInstanceId, annotation.getDisplayName())
+                            : "";
+                    row.append("," + value);
+                  });
+              fileContents.append(String.format(row + "\n"));
+            });
+
+    String projectId = tanagraExportConfiguration.getGcsBucketProjectId();
+    String bucketName = tanagraExportConfiguration.getGcsBucketName();
+    String fileName = "tanagra_export_annotations_" + System.currentTimeMillis();
+    if (StringUtils.isEmpty(projectId) || StringUtils.isEmpty(bucketName)) {
+      throw new SystemException(
```

Comment:
Would it be helpful to log `fileContents` here so it's not necessary to set up GCS for debugging purposes?

Reply:
Done